services/venice-controller/src/main/java/com/linkedin/venice/controller/StoreBackupVersionCleanupService.java

```diff
@@ -91,6 +125,83 @@ protected static boolean whetherStoreReadyToBeCleanup(Store store, long defaultB
   }
 
+  private boolean validateAllRouterOnCurrentVersion(Store store, String clusterName, int versionToValidate) {
+    List<String> urlList = admin.getHelixVeniceClusterResources(clusterName)
+        .getRoutersClusterManager()
+        .getLiveRouterInstances()
+        .stream()
+        .map(
+            instance -> urlMap
+                .computeIfAbsent(instance, k -> HelixUtils.getInstanceFromHelixInstanceName(k).getUrl(true)))
+        .collect(Collectors.toList());
+
+    for (String routerURL: urlList) {
+      try {
+        HttpGet routerRequest = new HttpGet(routerURL + "/" + TYPE_CURRENT_VERSION + "/" + store.getName());
+        HttpResponse response = storeClient.execute(routerRequest, null).get();
+        if (response.getStatusLine().getStatusCode() != HttpStatus.SC_OK) {
+          LOGGER.warn(
+              "Got status code {} from host {} while querying current version for store {}",
+              response.getStatusLine().getStatusCode(),
+              routerURL,
+              store.getName());
+          return false;
+        }
+        String responseBody;
+        try (InputStream bodyStream = response.getEntity().getContent()) {
+          responseBody = IOUtils.toString(bodyStream);
+        }
+        CurrentVersionResponse currentVersionResponse =
+            OBJECT_MAPPER.readValue(responseBody.getBytes(), CurrentVersionResponse.class);
+        if (currentVersionResponse.getCurrentVersion() != versionToValidate) {
+          return false;
+        }
+      } catch (Exception e) {
+        LOGGER.error("Got exception while getting current version for store {}", store.getName(), e);
+        return false;
+      }
+    }
+    return true;
+  }
+
+  private boolean validateAllServerOnCurrentVersion(Store store, String clusterName, int versionToValidate) {
+    Set<Instance> instances = admin.getLiveInstanceMonitor(clusterName).getAllLiveInstances();
+
+    for (Instance instance: instances) {
+      try {
+        HttpGet routerRequest = new HttpGet(
+            instance.getUrl() + "/" + QueryAction.METADATA.toString().toLowerCase() + "/" + store.getName());
+        HttpResponse response = storeClient.execute(routerRequest, null).get();
+        if (response.getStatusLine().getStatusCode() != HttpStatus.SC_OK) {
+          LOGGER.warn(
+              "Got status code {} from host {} while querying current version for store {}",
+              response.getStatusLine().getStatusCode(),
+              instance,
+              store.getName());
+          return false;
+        }
+        byte[] responseBody;
+        try (InputStream bodyStream = response.getEntity().getContent()) {
+          responseBody = IOUtils.toByteArray(bodyStream);
+        }
+        RecordDeserializer<MetadataResponseRecord> metadataResponseRecordRecordDeserializer =
+            SerializerDeserializerFactory.getAvroGenericDeserializer(MetadataResponseRecord.SCHEMA$);
```

Comment:
`MetadataResponseRecord` is evolvable, right?
Does the response header contain the schema id?

Reply:
To make it simple, maybe we can expose a new endpoint in server as well as the one in Router and it will be in json format instead of Avro.